### PR TITLE
Make DiscoveryNodes behave like a collection

### PR DESCRIPTION
**Pull request by arteam**
Extend from `AbstractCollection`. Advantages:

* Optimized version of `spliterator` based on the collection size which is implemented in `AbstractCollection` by default.
* Exposed `stream` method, so consumers don't need to call `StreamSupport.stream(discoveryNodes.spliterator(), false)`

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 1]
```diff
@@ -84,6 +84,16 @@ private DiscoveryNodes(
...
 
+    @Override
+    public int size() {
+        return nodes.size();
+    }
+
+    @Override
+    public Spliterator<DiscoveryNode> spliterator() {
+        return super.spliterator();
+    }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:95]: Is it intentional?

**Review by idegtiarenko — APPROVED** [review]
:+1:

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 1 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:95]: Good catch. It was used to find out all the usages of the spliterator, I will remove the override.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

**Pull request merged**